### PR TITLE
Add `unnecessary_clippy_cfg` lint

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Follow-up of https://github.com/rust-lang/rust-clippy/pull/12292.

r? @flip1995 

changelog: Add `unnecessary_clippy_cfg` lint

### CODE REVIEW

File: clippy_lints/src/attrs.rs [hunk 1]
```diff
@@ -958,6 +985,69 @@ fn check_deprecated_cfg_attr(cx: &EarlyContext<'_>, attr: &Attribute, msrv: &Msr
...
+
+fn check_clippy_cfg_attr(
+    cx: &EarlyContext<'_>,
+    cfg_attr: &rustc_ast::MetaItem,
+    behind_cfg_attr: &rustc_ast::MetaItem,
+    attr: &Attribute,
+) {
+    if cfg_attr.has_name(sym::clippy)
+        && let Some(ident) = behind_cfg_attr.ident()
+        && ["allow", "warn", "deny"].contains(&ident.name.as_str())
```
Comment (flip1995) [inline clippy_lints/src/attrs.rs:1003]: Also `forbid` and `expect` in the future. Is there some list of those words, that we can get from `rustc`?

File: tests/ui/unneedeed_clippy_cfg_attr.rs
Comment (flip1995) [inline tests/ui/unneedeed_clippy_cfg_attr.rs:1]: There's no stderr file for this test :thinking:

File: clippy_lints/src/attrs.rs [hunk 2]
```diff
@@ -958,6 +985,69 @@ fn check_deprecated_cfg_attr(cx: &EarlyContext<'_>, attr: &Attribute, msrv: &Msr
...
+                clippy_lints.push(item.span());
+            }
+        }
+        if nb_items == clippy_lints.len() {
+            if let Some(snippet) = snippet_opt(cx, behind_cfg_attr.span) {
+                span_lint_and_sugg(
+                    cx,
+                    UNNEEDED_CLIPPY_CFG,
+                    attr.span,
+                    "No need to put clippy lints behind a `clippy` cfg",
```
Comment (flip1995) [inline clippy_lints/src/attrs.rs:1022]:
```suggestion
                    "unnecessary `cfg_attr` to set clippy lint level",
```

No uppercase at the start of the error message. Same below.

File: clippy_lints/src/attrs.rs (hunk 1 above)
Comment (Urgau) [reply clippy_lints/src/attrs.rs:1003]: You could probably use [`Level::from_attr`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint_defs/enum.Level.html#method.from_attr) or [`Level::from_str`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint_defs/enum.Level.html#method.from_str).

File: clippy_lints/src/attrs.rs [hunk 3]
```diff
@@ -459,6 +461,30 @@ declare_clippy_lint! {
...
+    /// run by anything else than clippy.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// #![cfg_attr(clippy, allow(clippy::deprecated_cfg_attr))]
+    /// ```
+    ///
+    /// Use instead:
+    /// ```no_run
+    /// #![cfg_attr(clippy, allow(clippy::deprecated_cfg_attr))]
```
Comment (Urgau) [inline clippy_lints/src/attrs.rs:480]:
```suggestion
    /// #![allow(clippy::deprecated_cfg_attr)]
```

File: clippy_lints/src/attrs.rs [hunk 4]
```diff
@@ -459,6 +461,30 @@ declare_clippy_lint! {
...
+    /// ```
+    #[clippy::version = "1.78.0"]
+    pub UNNEEDED_CLIPPY_CFG,
```
Comment (Urgau) [inline clippy_lints/src/attrs.rs:483]: nit:
```suggestion
    pub UNNECESSARY_CLIPPY_CFG,
```

**Comment by GuillaumeGomez** [comment]
Fixed CI.

File: clippy_lints/src/attrs.rs [hunk 5]
```diff
@@ -459,6 +461,30 @@ declare_clippy_lint! {
     "usage of `cfg(feature = \"cargo-clippy\")` instead of `cfg(clippy)`"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `#[cfg_attr(clippy, allow(clippy::lint))]` and for
```
Comment (flip1995) [inline clippy_lints/src/attrs.rs:466]:
```suggestion
    /// Checks for `#[cfg_attr(clippy, allow(clippy::lint))]`
```

File: clippy_lints/src/attrs.rs [hunk 6]
```diff
@@ -958,6 +985,74 @@ fn check_deprecated_cfg_attr(cx: &EarlyContext<'_>, attr: &Attribute, msrv: &Msr
...
+fn check_clippy_cfg_attr(
+    cx: &EarlyContext<'_>,
+    cfg_attr: &rustc_ast::MetaItem,
+    behind_cfg_attr: &rustc_ast::MetaItem,
+    attr: &Attribute,
+) {
+    if cfg_attr.has_name(sym::clippy)
+        && let Some(ident) = behind_cfg_attr.ident()
+        // FIXME: replace with `from_symbol` once https://github.com/rust-lang/rust/pull/121230
+        // is merged.
```
Comment (flip1995) [inline clippy_lints/src/attrs.rs:1004]: The PR got merged. I will update this during the sync today.

File: tests/ui/unnecessary_clippy_cfg.rs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+//@no-rustfix
+
+#![warn(clippy::unnecessary_clippy_cfg)]
+#![cfg_attr(clippy, deny(clippy::non_minimal_cfg))]
+#![cfg_attr(clippy, deny(dead_code, clippy::non_minimal_cfg))]
+#![cfg_attr(clippy, deny(dead_code, clippy::non_minimal_cfg, clippy::maybe_misused_cfg))]
+
+#[cfg_attr(clippy, deny(clippy::non_minimal_cfg))]
+#[cfg_attr(clippy, deny(dead_code, clippy::non_minimal_cfg))]
+#[cfg_attr(clippy, deny(dead_code, clippy::non_minimal_cfg, clippy::maybe_misused_cfg))]
```
Comment (flip1995) [inline tests/ui/unnecessary_clippy_cfg.rs:10]: There's one test case missing with multiple Clippy lints, but no rustc lints.

**Review by flip1995 — COMMENTED** [review]
LGTM, 2 NITs left.

**Comment by GuillaumeGomez** [comment]
Fix nit and add new test case as suggested.

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Pull request merged**